File: secondWeek/SortList.java
class Solution {
    public ListNode sortList(ListNode head) {
        return quickSort(head, null);
    }
    
    private ListNode quickSort(ListNode startNode, ListNode endNode) {
        if (startNode == endNode) {
            return startNode;
        }
        
        int pivotValue = startNode.val;
        ListNode front = startNode;
        ListNode back = startNode;
        while (front != endNode) {
            if (front.val < pivotValue) {
                back = back.next;
                swapNodeValue(front, back);
            }
            front = front.next;
        }
        swapNodeValue(startNode, back);
        
        quickSort(startNode, back);
        quickSort(back.next, endNode);
        
        return startNode;
    }
    
    private void swapNodeValue(ListNode nodeOne, ListNode nodeTwo) {
        int value = nodeOne.val;
        nodeOne.val = nodeTwo.val;
        nodeTwo.val = value;
    }
}

File: thirtiethWeek/sort_list.py
class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head

        slow = fast = head
        while fast.next and fast.next.next:
            slow, fast = slow.next, fast.next.next
        slow.next, slow = None, slow.next
        p, q = self.sortList(head), self.sortList(slow)

        head = p if p.val < q.val else q
        while p and q:
            if p.val < q.val:
                while p.next and p.next.val < q.val:
                    p = p.next
                p.next, p = q, p.next
            else:
                while q.next and q.next.val < p.val:
                    q = q.next
                q.next, q = p, q.next
        return head
